website/src/middlewares/api-v3/getUserLanguage.js

```diff
@@ -0,0 +1,85 @@
+import { model as User } from '../../models/user';
+import accepts from 'accepts';
+import _ from 'lodash';
+import {
+  translations,
+  defaultLangCodes,
+  multipleVersionsLanguages,
+} from '../../libs/api-v3/i18n';
+
+function _getFromBrowser (req) {
+  let acceptedLanguages = accepts(req).languages();
+
+  let acceptable = _(acceptedLanguages).map((lang) => {
+    return lang.slice(0, 2);
+  }).uniq().value();
+
+  let matches = _.intersection(acceptable, defaultLangCodes);
+
+  let iAcceptedCompleteLang = matches.length > 0 ? multipleVersionsLanguages.indexOf(matches[0].toLowerCase()) : -1;
+
+  if (iAcceptedCompleteLang !== -1) {
+    let acceptedCompleteLang = _.find(acceptedLanguages, (accepted) => {
+      return accepted.slice(0, 2) === multipleVersionsLanguages[iAcceptedCompleteLang];
+    });
+
+    if (acceptedCompleteLang) {
+      acceptedCompleteLang = acceptedCompleteLang.toLowerCase();
+    } else {
+      return 'en';
+    }
+
+    if (matches[0] === 'es') {
+      // In case of a Latin American version of Spanish use 'es_419'
+      return multipleVersionsLanguages.es.indexOf(acceptedCompleteLang !== -1) ? 'es_419' : 'es';
+    } else if (matches[0] === 'zh') {
+      let iChinese = multipleVersionsLanguages.zh.indexOf(acceptedCompleteLang.toLowerCase());
+
+      return iChinese !== -1 ? multipleVersionsLanguages.zh[iChinese] : 'zh';
+    } else {
+      return 'en';
+    }
+  } else if (matches.length > 0) {
+    return matches[0].toLowerCase();
+  } else {
+    return 'en';
+  }
+}
+
+function _getFromUser (user, req) {
```

Comment:
yeah your code seems cleaner. About the if, i don't know, we can add the extra check but it isn't really necessary unless someone really messes with their data in a way that we don't allow. But definitively better than crashing everything. So yeah let's add it.

Actually for the code, something like this is even better:

```
let preferredLang = user && user.preferences && user.preferences.language;
let lang = translations[preferredLang] ? preferredLang : _getFromBrowser(req);

return lang;

```